File: scripts/releaseScript.mjs
import { promises as fs } from 'fs'
import path from 'path'

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import readlineSync from 'readline-sync'
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import pkg from 'shelljs'

const { cd, exec, error: shellError } = pkg
// import * as shell from 'shelljs'

// Import the package.json file
import packageFile from '../package.json' with { type: 'json' }
// import file from '../package.json' assert { type: 'json' }

process.env.FORCE_COLOR = '1'

function getVersion() {
  const packageJson = packageFile
  const version = packageJson.version
  return version
}

/**
 * Update the version number in package.json
 *
 * @param {string} newVersion
 * @param {string} projectDir
 */
async function updatePackageJsonVersion(newVersion, projectDir) {
  const packageJsonPath = path.join(projectDir, 'package.json')
  const packageJson = JSON.parse(await fs.readFile(packageJsonPath, 'utf8'))
  packageJson.version = newVersion
  await fs.writeFile(packageJsonPath, JSON.stringify(packageJson, null, 2) + '\n', 'utf8')
}

/**
 * Function to check last command's exit status
 *
 * @param {string} task
 */
function checkStatus(task) {
  if (shellError()) {
    console.log('\n-------------------------------------------------------------')
    console.error(`${task} failed, aborting the script.`)
    console.log('-------------------------------------------------------------\n')
    process.exit(1)
  }
}

/**
 * @param {string} projectPath
 */
async function attemptRelease(projectPath) {
  // Get the current version
  const currentVersion = getVersion()
  checkStatus('Get Current Version')
  // Change to the hosting directory or exit if it fails
  const hostingPath = path.join(projectPath, 'hosting')
  cd(hostingPath)
  checkStatus('Change Directory to Hosting')
  // Run linting
  console.log(
    '\n-------------------------------------------------------------\n' +
      'Linting...' +
      '\n-------------------------------------------------------------\n',
  )
  exec('yarn lint:prod')
  checkStatus('Lint')
  cd(projectPath)
  console.log(
    '\n-------------------------------------------------------------\n\n' + `Current Version: ${currentVersion}`,
  )
  let newVersion = currentVersion
  let gitPorcelain = false
  let tag = false
  if (exec('git status --porcelain', { silent: true }).stdout === '') {
    gitPorcelain = true
    console.log(
      '\n-------------------------------------------------------------\n' +
        'Git Repo is up to date, moving on...' +
        '\n-------------------------------------------------------------\n',
    )
  } else {
    console.log('\n-------------------------------------------------------------\n')
    // const incrementVersion = readlineSync.keyInYNStrict('Do you want to increment the version number?')
    const incrementVersion = readlineSync.keyIn(
      '>>> Do you want to increment the version number? ( yes [y], no [n], exit [x] ):  ',
      { limit: 'YyNncCxXqQ' },
    )
    if (['c', 'x', 'q'].includes(incrementVersion.toLowerCase())) {
      process.exit(0)
    }
    if (incrementVersion.toLowerCase() === 'y') {
      newVersion = readlineSync.question('>>> Enter the new version number:  ')
      if (/^[0-9]+?\.[0-9]+?\.[0-9]+$/.test(newVersion)) {
        await updatePackageJsonVersion(newVersion, projectPath)
        tag = true
        console.log(
          '\n-------------------------------------------------------------\n' +
            'Version update completed successfully.' +
            '\n-------------------------------------------------------------\n',
        )
      } else {
        console.error('\nInvalid version format. Version was not incremented.\n')
        process.exit(1)
      }
    }
  }
  checkStatus('Version Update')
  if (gitPorcelain) {
    console.log(
      '\n-------------------------------------------------------------\n' +
        'No changes to commit.' +
        '\n-------------------------------------------------------------\n',
    )
  } else {
    console.log(
      '\n-------------------------------------------------------------\n' +
        'Modified files:' +
        '\n-------------------------------------------------------------\n',
    )
    exec('git status --porcelain')
    checkStatus('Git Status')
    console.log('\n---\n')
    exec('git status')
    checkStatus('Git Status')
    readlineSync.question(
      '\n>>> Modified files shown above' + '\n>>> Press Enter to continue to the commit process, or Ctrl+C to abort.\n',
    )
    exec('git add .')
    checkStatus('Git Add')
    console.log(
      '\n-------------------------------------------------------------\n' +
        'Changes to be committed:' +
        '\n-------------------------------------------------------------\n',
    )
    exec('GIT_PAGER=cat git diff --cached')
    checkStatus('Git Diff')
    readlineSync.question(
      '\n>>> Diffs shown above' + '\n>>> Press Enter to continue to the commit process, or Ctrl+C to abort.\n',
    )
    console.log(
      '\n-------------------------------------------------------------\n' +
        'Starting Git commit process...' +
        '\n-------------------------------------------------------------\n',
    )
    const commitMsg = readlineSync.question('>>> Enter your commit message:  ')
    exec(`git commit -m "${commitMsg}"`)
    checkStatus('Git Commit')
  }
  if (tag) {
    console.log('\n-------------------------------------------------------------\n')
    // const tagCommit = readlineSync.keyInYNStrict('Do you want to TAG this commit with the version number?')
    const tagCommit = readlineSync.keyIn(
      '>>> Do you want to TAG this commit with the version number? ( yes [y], no [n], exit [x] ):  ',
      { limit: 'YyNncCxXqQ' },
    )
    if (['c', 'x', 'q'].includes(tagCommit.toLowerCase())) {
      process.exit(0)
    }
    if (tagCommit) {
      exec(`git tag -a "v${newVersion}" -m "Version ${newVersion}"`)
      console.log(
        '\n-------------------------------------------------------------\n' +
          `Tag v${newVersion} added.` +
          '\n-------------------------------------------------------------\n',
      )
    }
  }
  checkStatus('Tag Release')
  cd(hostingPath)
  console.log(
    '\n-------------------------------------------------------------\n' +
      'Building project...' +
      '\n-------------------------------------------------------------\n',
  )
  exec('yarn build:prod')
  checkStatus('Build')
  cd(projectPath)
  console.log(
    '\n-------------------------------------------------------------\n' +
      'Deploying project...' +
      '\n-------------------------------------------------------------\n',
  )
  exec('yarn deploy-rules')
  checkStatus('Deploy Rules')
  exec('yarn deploy')
  checkStatus('Deploy Hosting')
  console.log(
    '\n-------------------------------------------------------------\n' +
      'Script completed successfully.' +
      '\n-------------------------------------------------------------\n',
  )
}

/**
 * @param {string} projectPath
 */
function main(projectPath) {
  if (projectPath) {
    attemptRelease(projectPath).then(
      () => {
        process.exit(0)
      },
      (err) => {
        console.error(err)
        process.exit(1)
      },
    )
  } else {
    console.error('\nNo project path specified.\n')
    process.exit(1)
  }
}

main(process.argv[2])

File: scripts/release_script.py
import argparse
import json
import os
import re
import subprocess
import sys
from pathlib import Path
from typing import Optional, Union

# Set the environment variable for colored output
os.environ["FORCE_COLOR"] = "1"


def run_command(
    cmd: Union[list[str], str],
    cwd: Optional[Path] = None,
    timeout: Optional[int] = None,
    silent: bool = False,
) -> str:
    """
    Executes a shell command and returns its output.

    Args:
        cmd (Union[list[str], str]): The command to execute.
        cwd (Optional[Path], optional): The working directory. Defaults to None.
        timeout (Optional[int], optional): Timeout in seconds. Defaults to None.
        silent (bool, optional): If True, suppresses output. Defaults to False.

    Raises:
        RuntimeError: If the command execution fails.

    Returns:
        str: The standard output from the command.
    """
    if not silent:
        print(f"Running command: {cmd}")
    try:
        shell = isinstance(cmd, str)
        result = subprocess.run(
            cmd,
            capture_output=True,
            text=True,
            encoding="utf-8",
            check=True,
            cwd=cwd,
            shell=shell,
            timeout=timeout,
        )
        if not silent:
            print(result.stdout)
            if result.stderr:
                print("\n--ERROR--\n")
                print(result.stderr)
                print("\n-------\n")
        if result.returncode != 0:
            raise RuntimeError(f"Command failed with error: {result.stderr}")
        return result.stdout.strip()
    except subprocess.CalledProcessError as e:
        raise RuntimeError(f"Command failed with error: {e.stderr}")


def get_version(project_path: Path) -> str:
    """Retrieves the current version from package.json."""
    package_json_path = project_path / "package.json"
    with package_json_path.open("r", encoding="utf-8") as f:
        package_json = json.load(f)
    return package_json["version"]


def update_package_json_version(new_version: str, project_dir: Path) -> None:
    """Updates the version number in package.json."""
    package_json_path = project_dir / "package.json"
    with package_json_path.open("r", encoding="utf-8") as f:
        package_json = json.load(f)
    package_json["version"] = new_version
    with package_json_path.open("w", encoding="utf-8") as f:
        json.dump(package_json, f, indent=2)
        f.write("\n")


def attempt_release(project_path: Path) -> None:
    try:
        # Get the current version
        current_version = get_version(project_path)
        print("\n" + "-" * 61)
        print(f"Current Version: {current_version}")
        print("-" * 61 + "\n")
        new_version = current_version
        git_porcelain = False
        tag = False

        # Check Git status
        git_status_output = run_command(
            ["git", "status", "--porcelain"], cwd=project_path, silent=True
        )
        if git_status_output == "":
            git_porcelain = True
            print("-" * 61)
            print("Git Repo is up to date, moving on...")
            print("-" * 61 + "\n")
        else:
            print("\n" + "-" * 61 + "\n")
            while True:
                increment_version = input(
                    ">>> Do you want to increment the version number? ( yes [y], no [n], exit [x] ):  "
                ).strip().lower()
                if increment_version in ["c", "x", "q"]:
                    sys.exit(0)
                elif increment_version in ["y", "yes"]:
                    new_version = input(">>> Enter the new version number:  ").strip()
                    if re.match(r"^[0-9]+\.[0-9]+\.[0-9]+$", new_version):
                        update_package_json_version(new_version, project_path)
                        tag = True
                        print("\n" + "-" * 61)
                        print("Version update completed successfully.")
                        print("-" * 61 + "\n")
                        break
                    else:
                        print("\nInvalid version format. Version was not incremented.\n")
                        sys.exit(1)
                elif increment_version in ["n", "no"]:
                    break
                else:
                    print(
                        "Invalid input. Please enter yes [y], no [n], or exit [x].\n"
                    )

        # Linting
        print("\n" + "-" * 61)
        print("Linting...")
        print("-" * 61 + "\n")
        hosting_path = project_path / "hosting"
        run_command(["yarn", "lint:prod"], cwd=hosting_path)

        if git_porcelain:
            print("\n" + "-" * 61)
            print("No changes to commit.")
            print("-" * 61 + "\n")
        else:
            print("\n" + "-" * 61)
            print("Modified files:")
            print("-" * 61 + "\n")
            run_command(["git", "status", "--porcelain"], cwd=project_path)
            print("\n---\n")
            run_command(["git", "status"], cwd=project_path)
            input(
                "\n>>> Modified files shown above\n>>> Press Enter to continue to the commit process, or Ctrl+C to abort.\n"
            )
            run_command(["git", "add", "."], cwd=project_path)
            print("\n" + "-" * 61)
            print("Changes to be committed:")
            print("-" * 61 + "\n")
            env = os.environ.copy()
            env["GIT_PAGER"] = "cat"
            subprocess.run(
                ["git", "diff", "--cached"], cwd=project_path, env=env, check=True
            )
            input(
                "\n>>> Diffs shown above\n>>> Press Enter to continue to the commit process, or Ctrl+C to abort.\n"
            )
            print("\n" + "-" * 61)
            print("Starting Git commit process...")
            print("-" * 61 + "\n")
            commit_msg = input(">>> Enter your commit message:  ")
            run_command(["git", "commit", "-m", commit_msg], cwd=project_path)

        if tag:
            print("\n" + "-" * 61 + "\n")
            while True:
                tag_commit = input(
                    ">>> Do you want to TAG this commit with the version number? ( yes [y], no [n], exit [x] ):  "
                ).strip().lower()
                if tag_commit in ["c", "x", "q"]:
                    sys.exit(0)
                elif tag_commit in ["y", "yes"]:
                    run_command(
                        [
                            "git",
                            "tag",
                            "-a",
                            f"v{new_version}",
                            "-m",
                            f"Version {new_version}",
                        ],
                        cwd=project_path,
                    )
                    print("\n" + "-" * 61)
                    print(f"Tag v{new_version} added.")
                    print("-" * 61 + "\n")
                    break
                elif tag_commit in ["n", "no"]:
                    break
                else:
                    print(
                        "Invalid input. Please enter yes [y], no [n], or exit [x].\n"
                    )

        # Build the project
        print("\n" + "-" * 61)
        print("Building project...")
        print("-" * 61 + "\n")
        run_command(["yarn", "build:prod"], cwd=hosting_path)

        # Deploy the project
        print("\n" + "-" * 61)
        print("Deploying project...")
        print("-" * 61 + "\n")
        run_command(["yarn", "deploy-rules"], cwd=project_path)
        run_command(["yarn", "deploy"], cwd=project_path)

        print("\n" + "-" * 61)
        print("Script completed successfully.")
        print("-" * 61 + "\n")

    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)


def main():
    parser = argparse.ArgumentParser(description="Release script")
    parser.add_argument("project_path", type=str, help="Path to the project directory")
    args = parser.parse_args()

    project_path = Path(args.project_path).resolve()
    if not project_path.is_dir():
        print(f"Error: Project path {project_path} is not a directory.")
        sys.exit(1)

    attempt_release(project_path)


if __name__ == "__main__":
    main()
